sw/device/tests/crypto/cryptotest/firmware/aes.c

```diff
@@ -27,6 +32,312 @@ static const uint32_t kKeyMask[8] = {
+    return INTERNAL(status.value);
+  }
+
+  size_t ciphertext_bytes_written;
+  TRY(stream_gcm(&ctx, aad, plaintext, ciphertext, &ciphertext_bytes_written));
+  otcrypto_byte_buf_t remaining_ciphertext = {
+      .data = ciphertext.data + ciphertext_bytes_written,
+      .len = uj_data.input_length - ciphertext_bytes_written,
+  };
+  TRY(otcrypto_aes_gcm_encrypt_final(&ctx, tag_len, remaining_ciphertext,
+                                     &ciphertext_bytes_written, auth_tag));
+  memcpy(uj_output.stepwise_output, ciphertext.data, uj_output.output_length);
+
+  // Send digest to host via UART
+  RESP_OK(ujson_serialize_cryptotest_aes_gcm_output_t, uj, &uj_output);
+  return OK_STATUS(0);
+}
+
+status_t handle_aes_gcm_decryption(ujson_t *uj,
+                                   cryptotest_aes_gcm_data_t uj_data,
+                                   otcrypto_blinded_key_t key,
+                                   otcrypto_const_word32_buf_t iv,
+                                   otcrypto_const_byte_buf_t aad,
+                                   hardened_bool_t *success) {
+  // Text block length should be a multiple of `kAesBlockNumBytes`.
+  size_t nblocks = ceil_div(uj_data.input_length, kAesBlockNumBytes);
+  size_t text_buffer_length = nblocks * kAesBlockNumBytes;
+  
+  // Create buffer to store the input. For decryption, it is the ciphertext.
+  uint8_t input_buf[text_buffer_length];
+  memcpy(input_buf, uj_data.input, uj_data.input_length);
+
+  // Set the ciphertext field
+  otcrypto_const_byte_buf_t ciphertext = {
+      .data = input_buf,
+      .len = text_buffer_length,
+  };
+
+  // Set the plaintext field
+  uint8_t plaintext_data[text_buffer_length];
+  otcrypto_byte_buf_t plaintext = {
+      .data = plaintext_data,
+      .len = text_buffer_length,
+  };
+
+  // Set the auth_tag and tag_len
+  size_t tag_num_words = ceil_div(uj_data.tag_length, sizeof(uint32_t));
+  uint32_t tag_buf[tag_num_words];
+  memcpy(tag_buf, uj_data.tag, uj_data.tag_length);
+  otcrypto_const_word32_buf_t auth_tag = {
+      .data = tag_buf,
+      .len = tag_num_words,
+  };
+  otcrypto_aes_gcm_tag_len_t tag_len = get_tag_length(tag_num_words * sizeof(uint32_t));
+
+  // Test oneshot API
+  otcrypto_status_t status = otcrypto_aes_gcm_decrypt(
+      &key, ciphertext, iv, aad, tag_len, auth_tag, plaintext, success);
+
+  if (status.value != kOtcryptoStatusValueOk) {
+    return INTERNAL(status.value);
+  }
+
+  // Copy oneshot output to uJSON typer
+  cryptotest_aes_gcm_output_t uj_output;
+  uj_output.output_length = uj_data.input_length;
+  memcpy(uj_output.oneshot_output, plaintext.data, uj_output.output_length);
+
+  // Test stepwise API
+  otcrypto_aes_gcm_context_t ctx;
+  status = otcrypto_aes_gcm_decrypt_init(&key, iv, &ctx);
+  if (status.value != kOtcryptoStatusValueOk) {
+    return INTERNAL(status.value);
+  }
+
+  size_t plaintext_bytes_written;
+  TRY(stream_gcm(&ctx, aad, ciphertext, plaintext, &plaintext_bytes_written));
```

Comment:
When I wrote the test for stepwise SHA-2, @jadephilipoom suggested splitting the input into two parts to get better coverage of the intermediate call. Not sure if that would work with GCM.

Reply:
Tested with partition size 2 and looks like it works.

I noticed we split the AES-GCM test vectors into more than two parts (10 for AAD and 20 for input strings) in unit tests. I am okay going in any direction though.